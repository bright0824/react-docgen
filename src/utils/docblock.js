/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

/**
 * Helper functions to work with docblock comments.
 */

var DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gmi;

function parseDocblock(str) {
  var lines = str.split('\n');
  for (var i = 0, l = lines.length; i < l; i++) {
    lines[i] = lines[i].replace(/^\s*\*\s?/, '');
  }
  return lines.join('\n').trim();
}

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath): ?string {
  if (path.node.comments) {
    var comments = path.node.comments.filter(function(comment) {
      return comment.leading &&
        comment.type === 'CommentBlock' &&
        comment.value.indexOf('*\n') === 0;
    });
    if (comments.length > 0) {
      return parseDocblock(comments[comments.length - 1].value);
    }
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Object {
  var doclets = Object.create(null);
  var match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
