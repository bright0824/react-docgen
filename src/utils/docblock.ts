/**
 * Helper functions to work with docblock comments.
 */

import { CommentKind } from 'ast-types/gen/kinds';
import type { NodePath } from 'ast-types/lib/node-path';

const DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gim;

function parseDocblock(str: string): string {
  // Does not use \s in the regex as this would match also \n and conflicts
  // with windows line endings.
  return str.replace(/^[ \t]*\*[ \t]?/gm, '').trim();
}

const DOCBLOCK_HEADER = /^\*\s/;

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath, trailing = false): string | null {
  let comments: CommentKind[] = [];
  if (trailing && path.node.trailingComments) {
    comments = path.node.trailingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.leadingComments) {
    comments = path.node.leadingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.comments) {
    comments = path.node.comments.filter(
      comment =>
        comment.leading &&
        comment.type === 'CommentBlock' &&
        DOCBLOCK_HEADER.test(comment.value),
    );
  }

  if (comments.length > 0) {
    return parseDocblock(comments[comments.length - 1].value);
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Record<string, unknown> {
  const doclets = Object.create(null);
  let match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
