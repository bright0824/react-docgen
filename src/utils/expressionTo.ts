/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';
import type { NodePath } from 'ast-types/lib/node-path';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): string[] {
  const parts = [path];
  let result: string[] = [];

  while (parts.length > 0) {
    path = parts.shift() as NodePath;
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        // @ts-ignore
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      // @ts-ignore
      result.push(node.raw);
      continue;
    } else if (t.FunctionExpression.check(node)) {
      result.push('<function>');
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        if (t.SpreadElement.check(property.node)) {
          return `...${toString(property.get('argument'), importer)}`;
        } else {
          return (
            toString(property.get('key'), importer) +
            ': ' +
            toString(property.get('value'), importer)
          );
        }
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
