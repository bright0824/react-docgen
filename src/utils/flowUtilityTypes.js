/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (types.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return Boolean(idPath) && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  return path.get('typeParameters', 'params', 0);
}
