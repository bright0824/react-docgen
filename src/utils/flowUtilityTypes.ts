import { namedTypes as t } from 'ast-types';
import type { NodePath } from 'ast-types/lib/node-path';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  while (isSupportedUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}
