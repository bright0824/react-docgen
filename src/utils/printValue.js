/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import recast from 'recast';

/**
 * Prints the given path without leading or trailing comments.
 */
export default function printValue(path: NodePath): string {
  if (path.node.comments) {
    path.node.comments.length = 0;
  }

  return recast.print(path).code;
}
