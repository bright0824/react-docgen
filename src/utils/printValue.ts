import type { NodePath } from 'ast-types/lib/node-path';
import strip from 'strip-indent';
import toBabel from 'estree-to-babel';
import generate from '@babel/generator';
import { FileNodeWithOptions } from '../babelParser';

function deindent(code: string): string {
  const firstNewLine = code.indexOf('\n');

  return (
    code.slice(0, firstNewLine + 1) +
    // remove indentation from all lines except first.
    strip(code.slice(firstNewLine + 1))
  );
}

function getSrcFromAst(path: NodePath): string {
  do {
    if (path.node.type === 'File') {
      return (path.node as FileNodeWithOptions).__src;
    }
    path = path.parentPath;
  } while (path != null);

  throw new Error('Could not find source attached to File node');
}

/**
 * Prints the given path without leading or trailing comments.
 */
export default function printValue(path: NodePath): string {
  if (path.node.start == null) {
    try {
      const nodeCopy = {
        ...path.node,
      };

      // `estree-to-babel` expects the `comments` property to exist on the top-level node
      if (!nodeCopy.comments) {
        nodeCopy.comments = [];
      }

      return generate(toBabel(nodeCopy), {
        comments: false,
        concise: true,
      }).code;
    } catch (err: any) {
      throw new Error(
        `Cannot print raw value for type '${path.node.type}'. Please report this with an example at https://github.com/reactjs/react-docgen/issues.

Original error:
${err.stack}`,
      );
    }
  }
  const src = getSrcFromAst(path);

  return deindent(src.slice(path.node.start, path.node.end));
}
