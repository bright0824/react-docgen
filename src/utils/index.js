/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import * as docblock from './docblock';

export {
  docblock,
};

export {default as getClassMemberValuePath} from './getClassMemberValuePath';
export {default as getMemberExpressionRoot} from './getMemberExpressionRoot';
export {default as getMembers} from './getMembers';
export {default as getMemberValuePath} from './getMemberValuePath';
export {default as getNameOrValue} from './getNameOrValue';
export {default as getPropertName} from './getPropertyName';
export {default as getPropertyValuePath} from './getPropertyValuePath';
export {default as getPropType} from './getPropType';
export {default as isExportsOrModuleAssignment} from './isExportsOrModuleAssignment';
export {default as isReactComponentClass} from './isReactComponentClass.js';
export {default as isReactCreateClassCall} from './isReactCreateClassCall';
export {default as isReactModuleName} from './isReactModuleName';
export {default as match} from './match';
export {default as normalizeClassDefiniton} from './normalizeClassDefinition';
export {default as printValue} from './printValue';
export {default as resolveExportDeclaration} from './resolveExportDeclaration';
export {default as resolveToModule} from './resolveToModule';
export {default as resolveToValue} from './resolveToValue';
