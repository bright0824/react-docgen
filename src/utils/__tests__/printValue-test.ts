import type { NodePath } from '@babel/traverse';
import type { ExpressionStatement } from '@babel/types';
import { parse } from '../../../tests/utils';
import printValue from '../printValue';

describe('printValue', () => {
  function pathFromSource(source: string): NodePath {
    return parse.statement<ExpressionStatement>(source).get('expression');
  }

  it('does not print leading comments', () => {
    expect(printValue(pathFromSource('//foo\nbar'))).toEqual('bar');
  });

  it('does not print trailing comments', () => {
    expect(printValue(pathFromSource('bar//foo'))).toEqual('bar');
  });

  it('deindents code', () => {
    expect(
      printValue(
        pathFromSource(`    (    function () {
      return x;
    })`),
      ),
    ).toMatchSnapshot();
  });
});
