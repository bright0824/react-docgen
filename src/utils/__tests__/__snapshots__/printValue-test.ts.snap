// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printValue deindents code 1`] = `
"function () {
  return x;
}"
`;
